src/ehttpc.erl

```diff
@@ -141,6 +141,15 @@ request(Worker, Method, Request, Timeout, Retry) when is_pid(Worker) ->
     try gen_server:call(Worker, ?REQ(Method, Request, ExpireAt), Timeout + 500) of
         %% gun will reply {gun_down, _Client, _, normal, _KilledStreams, _} message
         %% when connection closed by keepalive
+
+        %% If `Reason' = `normal', we should just retry without
+        %% consuming a retry credit, as it could be a race condition
+        %% where the gun process is down (e.g.: when the server closes
+        %% the connection), and then requests would be ignored while
+        %% the `gun' process is terminating.
+        {error, normal} ->
```

Comment:
It is hard to reason about where this `{error, normal}` comes from.
If it is not replied by the gun but by the `reply_error_for_sent_reqs/2` when we received `DOWN` while there are some pending requests, we'd better change the `normal` to `gun_down_before_sent`, and then unconditionally retry the request on `gun_down_before_sent` is returned.

Reply:
I also had some trouble provoking that failure in the unit tests.

From what I can tell, this `{error, normal}` can come from these two points:

- https://github.com/emqx/ehttpc/blob/aa4b26c73f7089820b7fad6696e6dde7f87285cf/src/ehttpc.erl#L280-L285
- https://github.com/emqx/ehttpc/blob/aa4b26c73f7089820b7fad6696e6dde7f87285cf/src/ehttpc.erl#L288-L291

In the first case, I think that `gun` will only send that `gun_down` message with `normal` reason when it's in some particular state(s), so that's probably why it's hard to reproduce in a controlled manner.  :thinking:

Reply:
The test case I added hits the `DOWN` case, but the `gun_down` one is trickier to trigger :thinking: